### ISSUE TITLE
Custom cache handler with dynamic route + static params always shows 404 page

### ISSUE THREAD

**Issue by maxquinn** [vercel/next.js#67771]
### Link to the code that reproduces this issue

https://github.com/maxquinn/test-cache-static-bug

### To Reproduce

1. `npm i`
2. `npm run build`
3. `npm start`
4. navigate to http://localhost:3000/test/should-work

### Current vs. Expected behavior

Navigating to http://localhost:3000/test/should-work should return the page but instead just shows the nextjs 404 page. You can see the page has been built statically during the `next build` step so it should exist. 

The page loads correctly if we disable the custom cache handler.

Regular pages also load fine. It only becomes an issue when we use `generateStaticParams` with `export const dynamicParams = false`

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:14:38 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6020
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 20.14.0
  npm: 10.7.0
  Yarn: N/A
  pnpm: 9.4.0
Relevant Packages:
  next: 14.2.5 // Latest available version is detected (14.2.5).
  eslint-config-next: N/A
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Runtime

### Which stage(s) are affected? (Select all that apply)

next start (local)

**Comment by Aleesssino** [comment]
It works when you remove 
```
export const dynamicParams = false;
```
from app/test/[test]/page.tsx.

For similar errors, it typically works when deployed on Vercel, but removing this line provides a "safer solution."

**Comment by maxquinn** [comment]
@Aleesssino Yes I know it works when removing the dynamicParams route segment config but the point is that I don't want dynamic params at all. I have a set list of params so that the page can be static at build time. 

Also we are not using Vercel for our production deployment.
